### PR TITLE
Extending pre-operation to numbers represented as text

### PR DESCRIPTION
**Pull request by MISBMS**

### CODE REVIEW

File: src/main/java/io/appform/jsonrules/expressions/equality/EqualsExpression.java [hunk 1]
```diff
@@ -35,24 +35,46 @@
...
 
-    @Override
-    protected boolean evaluate(ExpressionEvaluationContext context, String path, JsonNode evaluatedNode) {
-        return !ComparisonUtils.isNodeMissingOrNull(evaluatedNode) && ComparisonUtils.compare(evaluatedNode, value) == 0;
-    }
+	@Override
+	protected boolean evaluate(ExpressionEvaluationContext context, String path, JsonNode evaluatedNode) {
+		if (extractValueFromPath) {
+			JsonNode jsonNode = context.getNode().at(String.valueOf(value));
+			if (jsonNode.isNumber()) {
```
Comment (santanusinha) [inline src/main/java/io/appform/jsonrules/expressions/equality/EqualsExpression.java]: What if node is not present?

File: src/main/java/io/appform/jsonrules/expressions/equality/EqualsExpression.java [hunk 2]
```diff
@@ -35,20 +35,46 @@
...
+	public EqualsExpression(String path, Object value, boolean extractValueFromPath, PreOperation<?> preoperation) {
+		this(path, value, extractValueFromPath, false, preoperation);
+	}
+
+	@Override
+	protected boolean evaluate(ExpressionEvaluationContext context, String path, JsonNode evaluatedNode) {
+		if (extractValueFromPath) {
+			JsonNode jsonNode = context.getNode().at(String.valueOf(value));
+			if (jsonNode.isNumber()) {
+				if (jsonNode.isIntegralNumber()) {
```
Comment (santanusinha) [inline src/main/java/io/appform/jsonrules/expressions/equality/EqualsExpression.java]: Move this section to ComparionUtils

File: src/main/java/io/appform/jsonrules/expressions/equality/EqualsExpression.java [hunk 3]
```diff
@@ -35,20 +35,46 @@
...
+					return !ComparisonUtils.isNodeMissingOrNull(evaluatedNode)
```
Comment (santanusinha) [inline src/main/java/io/appform/jsonrules/expressions/equality/EqualsExpression.java]: isNodeMissingOrNull can be checked once globally before going into the decision tree?

File: src/main/java/io/appform/jsonrules/expressions/equality/NotEqualsExpression.java [hunk 4]
```diff
@@ -36,23 +33,41 @@
...
+    public NotEqualsExpression(String path, Object value, boolean extractValueFromPath, Boolean defaultResult, PreOperation<?> preoperation) {
+        super(ExpressionType.not_equals, path, ComparisonUtils.getDefaultResult(defaultResult, true), preoperation);
         this.value = value;
+        this.extractValueFromPath = extractValueFromPath;
     }
 
     @Override
     protected boolean evaluate(ExpressionEvaluationContext context, String path, JsonNode evaluatedNode) {
+    	if (extractValueFromPath) {
+			JsonNode jsonNode = context.getNode().at(String.valueOf(value));
```
Comment (santanusinha) [inline src/main/java/io/appform/jsonrules/expressions/equality/NotEqualsExpression.java]: Are you using 4 spaces as tab? If not please change it to that.

File: src/main/java/io/appform/jsonrules/expressions/equality/NotEqualsExpression.java [hunk 5]
```diff
@@ -36,23 +33,41 @@
...
+			if (jsonNode.isNumber()) {
```
Comment (santanusinha) [inline src/main/java/io/appform/jsonrules/expressions/equality/NotEqualsExpression.java]: Please check comments above. Same code. Needs to be refactored.

File: src/main/java/io/appform/jsonrules/expressions/numeric/NumericJsonPathBasedExpression.java [hunk 6]
```diff
@@ -33,31 +33,48 @@
...
+	@Override
+	protected final boolean evaluate(ExpressionEvaluationContext context, String path, JsonNode evaluatedNode) {
+		if (null == evaluatedNode || !evaluatedNode.isNumber()) {
+			return false;
+		}
+
+		Number numericalValue;
+		if (extractValueFromPath) {
+			JsonNode jsonNode = context.getNode().at(String.valueOf(value));
+			if (jsonNode.isIntegralNumber()) {
```
Comment (santanusinha) [inline src/main/java/io/appform/jsonrules/expressions/numeric/NumericJsonPathBasedExpression.java]: Check if node is present

File: src/main/java/io/appform/jsonrules/expressions/string/EndsWithExpression.java [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+	@Builder
+	public EndsWithExpression(String path, String value, boolean ignoreCase, boolean extractValueFromPath,
+			Boolean defaultResult, PreOperation<?> preoperation) {
+		super(ExpressionType.ends_with, path, value, ignoreCase, extractValueFromPath, defaultResult, preoperation);
+	}
+
+	@Override
+	protected boolean evaluate(String leftValue, String rightValue, boolean ignoreCase) {
+		if (null != leftValue) {
+			if (ignoreCase) {
```
Comment (santanusinha) [inline src/main/java/io/appform/jsonrules/expressions/string/EndsWithExpression.java]: Check and fix tabs

File: src/main/java/io/appform/jsonrules/expressions/string/EndsWithExpression.java [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+	@Builder
+	public EndsWithExpression(String path, String value, boolean ignoreCase, boolean extractValueFromPath,
+			Boolean defaultResult, PreOperation<?> preoperation) {
+		super(ExpressionType.ends_with, path, value, ignoreCase, extractValueFromPath, defaultResult, preoperation);
+	}
+
+	@Override
+	protected boolean evaluate(String leftValue, String rightValue, boolean ignoreCase) {
+		if (null != leftValue) {
```
Comment (santanusinha) [inline src/main/java/io/appform/jsonrules/expressions/string/EndsWithExpression.java]: Strings.isNullOrEmpty everywhere. Check and fix in other places as well.

File: src/main/java/io/appform/jsonrules/expressions/string/StringJsonPathBasedExpression.java [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+        this.extractValueFromPath = extractValueFromPath;
+    }
+
+    @Override
+    protected final boolean evaluate(ExpressionEvaluationContext context, String path, JsonNode evaluatedNode) {
+    	 if(!evaluatedNode.isTextual()) {
+             return false;
+         }
+    	 if (extractValueFromPath) {
+    		 final String extractedValue = context.getNode().at(value).asText();
```
Comment (santanusinha) [inline src/main/java/io/appform/jsonrules/expressions/string/StringJsonPathBasedExpression.java]: Will cause NPE if there is no node in this path. Fix this in all the places and introduce testcases.

File: src/main/java/io/appform/jsonrules/expressions/equality/EqualsExpression.java (hunk 2 above)
Comment (MISBMS) [reply src/main/java/io/appform/jsonrules/expressions/equality/EqualsExpression.java]: Done

File: src/main/java/io/appform/jsonrules/expressions/equality/EqualsExpression.java (hunk 3 above)
Comment (MISBMS) [reply src/main/java/io/appform/jsonrules/expressions/equality/EqualsExpression.java]: Done

File: src/main/java/io/appform/jsonrules/expressions/equality/NotEqualsExpression.java (hunk 4 above)
Comment (MISBMS) [reply src/main/java/io/appform/jsonrules/expressions/equality/NotEqualsExpression.java]: Done, at all places

File: src/main/java/io/appform/jsonrules/expressions/string/EndsWithExpression.java (hunk 7 above)
Comment (MISBMS) [reply src/main/java/io/appform/jsonrules/expressions/string/EndsWithExpression.java]: Done

File: src/main/java/io/appform/jsonrules/expressions/string/EndsWithExpression.java (hunk 8 above)
Comment (MISBMS) [reply src/main/java/io/appform/jsonrules/expressions/string/EndsWithExpression.java]: Done

File: src/main/java/io/appform/jsonrules/expressions/string/StringJsonPathBasedExpression.java (hunk 9 above)
Comment (MISBMS) [reply src/main/java/io/appform/jsonrules/expressions/string/StringJsonPathBasedExpression.java]: Done

File: src/main/java/io/appform/jsonrules/expressions/numeric/NumericJsonPathBasedExpression.java (hunk 6 above)
Comment (MISBMS) [reply src/main/java/io/appform/jsonrules/expressions/numeric/NumericJsonPathBasedExpression.java]: Done

File: src/main/java/io/appform/jsonrules/expressions/equality/EqualsExpression.java (hunk 1 above)
Comment (MISBMS) [reply src/main/java/io/appform/jsonrules/expressions/equality/EqualsExpression.java]: Handled at the very beginning.

**Pull request merged**